query-algebrizer/src/cc.rs

```diff
@@ -943,6 +979,124 @@ mod testing {
         ]);
     }
 
+    /// This test ensures that we do less work if we know the attribute thanks to a var lookup.
+    #[test]
+    fn test_apply_unattributed_but_bound_pattern_with_returned() {
+        let mut cc = ConjoiningClauses::default();
+        let mut schema = Schema::default();
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "bar"), 99);
+        add_attribute(&mut schema, 99, Attribute {
+            value_type: ValueType::Boolean,
+            ..Default::default()
+        });
+
+        let x = Variable(PlainSymbol::new("?x"));
+        let a = Variable(PlainSymbol::new("?a"));
+        let v = Variable(PlainSymbol::new("?v"));
+
+        cc.input_variables.insert(a.clone());
+        cc.value_bindings.insert(a.clone(), TypedValue::Keyword(NamespacedKeyword::new("foo", "bar")));
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(x.clone()),
+            attribute: PatternNonValuePlace::Variable(a.clone()),
+            value: PatternValuePlace::Variable(v.clone()),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        // println!("{:#?}", cc);
+
+        let d0_e = QualifiedAlias("datoms00".to_string(), DatomsColumn::Entity);
+        let d0_a = QualifiedAlias("datoms00".to_string(), DatomsColumn::Attribute);
+        let d0_v = QualifiedAlias("datoms00".to_string(), DatomsColumn::Value);
+
+        assert!(!cc.is_known_empty);
+        assert_eq!(cc.from, vec![SourceAlias(DatomsTable::Datoms, "datoms00".to_string())]);
+
+        // ?x must be a ref.
+        assert_eq!(cc.known_types.get(&x).unwrap(), &ValueType::Ref);
+
+        // ?x is bound to datoms0.e.
+        assert_eq!(cc.column_bindings.get(&x).unwrap(), &vec![d0_e.clone()]);
+        assert_eq!(cc.wheres, vec![
+                   ColumnConstraint::EqualsEntity(d0_a, 99),
+        ]);
```

Comment:
No. Remember, this knife cuts both ways: now that we've bound `?a` to `:foo/bar`, which we know has a range of `boolean`, any value in `v` with that `a` must be a boolean. `1` can never mean anything but `true` if `a=99`.

This is the main reason we don't have a forest of type tag checks and tag projections in every query: whenever we know the attribute, we can safely and statically assume the type of each value.